src/meta/src/stream/stream_manager.rs

```diff
@@ -105,6 +110,136 @@ where
         })
     }
 
+    #[async_recursion]
+    async fn resolve_chain_node_inner(
+        &self,
+        stream_node: &mut StreamNode,
+        actor_id: ActorId,
+        locations: &ScheduledLocations,
+        upstream_parallel_unit_info: &HashMap<TableId, BTreeMap<ParallelUnitId, ActorId>>,
+        ctx: &mut CreateMaterializedViewContext,
+    ) -> Result<()> {
+        // if node is chain node, we insert upstream ids into chain's input(merge)
+        if let Some(Node::ChainNode(ref mut chain)) = stream_node.node {
+            // get upstream table id
+            let table_id = TableId::from(&chain.table_ref_id);
+
+            let (upstream_actor_id, parallel_unit_id) = {
+                // 1. use table id to get upstream parallel_unit->actor_id mapping
+                let upstream_parallel_actor_mapping =
+                    upstream_parallel_unit_info.get(&table_id).unwrap();
+                // 2. use our actor id to get our parallel unit id
+                let parallel_unit_id = locations.actor_locations.get(&actor_id).unwrap().id;
+                // 3. and use our parallel unit id to get upstream actor id
+                (
+                    upstream_parallel_actor_mapping
+                        .get(&parallel_unit_id)
+                        .unwrap(),
+                    parallel_unit_id,
+                )
+            };
+
+            // fill upstream node-actor info for later use
+            let chain_upstream_table_node_actors =
+                self.fragment_manager.table_node_actors(&table_id).await?;
```

Comment:
Better to resolve this info ahead of time? So we can resolve #1188 as well. :)

Reply:
I moved this out of the inner recursion. Actually we don't have blocking interfaces now, but there's some code redundency since we have two `resolve_chain_node` now 😄 .